//cmd/kubeadm fails to build in pure Go mode under Bazel due to cadvisor dependencies

**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

I've been working on updating our `rules_go` dependency to be able to build our "static" binaries with cgo disabled to match the output from the non-bazel build.

It almost works, but `//cmd/kubeadm` is failing to build because it's trying to link in an empty library:
```
ERROR: .../code/kubernetes/src/k8s.io/kubernetes/vendor/github.com/google/cadvisor/container/raw/BUILD:3:1: GoCompile vendor/github.com/google/cadvisor/container/raw/linux_amd64_pure_stripped/go_default_library~/github.com/google/cadvisor/container/raw.a failed (Exit 1): compile failed: error executing command
  (cd .../.cache/bazel/_bazel_jgrafton/75e5737b0f848bf1b7c3ab3303c8bf72/execroot/__main__ && \
  exec env - \
  bazel-out/host/bin/external/io_bazel_rules_go/go/tools/builders/compile -go bazel-out/host/bin/external/go_stdlib_linux_amd64_pure/bin/go -root_file bazel-out/host/bin/external/go_stdlib_linux_amd64_pure/ROOT -goos linux -goarch amd64 '-cgo=0' -package_list external/go_sdk/packages.txt -src vendor/github.com/google/cadvisor/container/raw/factory.go -src vendor/github.com/google/cadvisor/container/raw/handler.go -dep github.com/golang/glog -dep github.com/google/cadvisor/container -dep github.com/google/cadvisor/container/common -dep github.com/google/cadvisor/container/libcontainer -dep github.com/google/cadvisor/fs -dep github.com/google/cadvisor/info/v1 -dep github.com/google/cadvisor/machine -dep github.com/google/cadvisor/manager/watcher -dep github.com/opencontainers/runc/libcontainer/cgroups -dep github.com/opencontainers/runc/libcontainer/cgroups/fs -dep github.com/opencontainers/runc/libcontainer/configs -I 'bazel-out/k8-fastbuild/bin/vendor/github.com/golang/glog/linux_amd64_pure_stripped/go_default_library~/' -I 'bazel-out/k8-fastbuild/bin/vendor/github.com/google/cadvisor/container/linux_amd64_pure_stripped/go_default_library~/' -I 'bazel-out/k8-fastbuild/bin/vendor/github.com/google/cadvisor/container/common/linux_amd64_pure_stripped/go_default_library~/' -I 'bazel-out/k8-fastbuild/bin/vendor/github.com/google/cadvisor/container/libcontainer/linux_amd64_pure_stripped/go_default_library~/' -I 'bazel-out/k8-fastbuild/bin/vendor/github.com/google/cadvisor/fs/linux_amd64_pure_stripped/go_default_library~/' -I 'bazel-out/k8-fastbuild/bin/vendor/github.com/google/cadvisor/info/v1/linux_amd64_pure_stripped/go_default_library~/' -I 'bazel-out/k8-fastbuild/bin/vendor/github.com/google/cadvisor/machine/linux_amd64_pure_stripped/go_default_library~/' -I 'bazel-out/k8-fastbuild/bin/vendor/github.com/google/cadvisor/manager/watcher/linux_amd64_pure_stripped/go_default_library~/' -I 'bazel-out/k8-fastbuild/bin/vendor/github.com/opencontainers/runc/libcontainer/cgroups/linux_amd64_pure_stripped/go_default_library~/' -I 'bazel-out/k8-fastbuild/bin/vendor/github.com/opencontainers/runc/libcontainer/cgroups/fs/linux_amd64_pure_stripped/go_default_library~/' -I 'bazel-out/k8-fastbuild/bin/vendor/github.com/opencontainers/runc/libcontainer/configs/linux_amd64_pure_stripped/go_default_library~/' -o 'bazel-out/k8-fastbuild/bin/vendor/github.com/google/cadvisor/container/raw/linux_amd64_pure_stripped/go_default_library~/github.com/google/cadvisor/container/raw.a' -trimpath . -I . -- -p github.com/google/cadvisor/container/raw)

Use --sandbox_debug to see verbose messages from the sandbox
vendor/github.com/google/cadvisor/container/raw/factory.go:23:2: import .../.cache/bazel/_bazel_jgrafton/75e5737b0f848bf1b7c3ab3303c8bf72/bazel-sandbox/4839007188253519489/execroot/__main__/bazel-out/k8-fastbuild/bin/vendor/github.com/google/cadvisor/container/libcontainer/linux_amd64_pure_stripped/go_default_library~/github.com/google/cadvisor/container/libcontainer.a: reading input: EOF
2017/11/28 15:05:07 error running compiler: exit status 1
Target //cmd/kubeadm:kubeadm failed to build
```

The issue seems to be that `cmd/kubeadm` depends on `pkg/kubelet/apis/kubeletconfig/validation`, which in turn depends on `pkg/kubelet/cm` for a few constants.

`pkg/kubelet/cm` is a fairly large package, and it has has build-tag controlled cgo and !cgo variants. The cgo variant depends on cadvisor, libcontainer, and other such cgo-heavy libraries, whereas the !cgo variants do not.

`go build` is pretty lazy about compiling dependencies, so it never builds cadvisor, libcontainer, etc in pure mode. Bazel doesn't have this ability, however, and so it tries to build these libraries in pure mode and fails.

Trying to get these end libraries to build in pure mode seems futile, especially since nobody would ever use them that way. A more viable fix is to break out the constants needed by `pkg/kubelet/apis/kubeletconfig/validation` into a separate package, which simplifies the build and prevents us from ever depending on these cgo-heavy libraries.

I'm not sure if there's precedence for this sort of refactoring, or where exactly the constants should go.

/area build-release

--- a/pkg/kubelet/apis/kubeletconfig/validation/BUILD
+++ b/pkg/kubelet/apis/kubeletconfig/validation/BUILD
@@ -12,7 +12,7 @@ go_library(
     importpath = "k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig/validation",
     deps = [
         "//pkg/kubelet/apis/kubeletconfig:go_default_library",
-        "//pkg/kubelet/cm:go_default_library",
+        "//pkg/kubelet/types:go_default_library",
         "//vendor/k8s.io/apimachinery/pkg/util/errors:go_default_library",
         "//vendor/k8s.io/apimachinery/pkg/util/validation:go_default_library",
     ],

--- a/pkg/kubelet/apis/kubeletconfig/validation/validation.go
+++ b/pkg/kubelet/apis/kubeletconfig/validation/validation.go
@@ -22,7 +22,7 @@ import (
 	utilerrors "k8s.io/apimachinery/pkg/util/errors"
 	utilvalidation "k8s.io/apimachinery/pkg/util/validation"
 	"k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig"
-	containermanager "k8s.io/kubernetes/pkg/kubelet/cm"
+	kubetypes "k8s.io/kubernetes/pkg/kubelet/types"
 )
 
 // ValidateKubeletConfiguration validates `kc` and returns an error if it is invalid
@@ -91,13 +91,13 @@ func ValidateKubeletConfiguration(kc *kubeletconfig.KubeletConfiguration) error 
 	}
 	for _, val := range kc.EnforceNodeAllocatable {
 		switch val {
-		case containermanager.NodeAllocatableEnforcementKey:
-		case containermanager.SystemReservedEnforcementKey:
-		case containermanager.KubeReservedEnforcementKey:
+		case kubetypes.NodeAllocatableEnforcementKey:
+		case kubetypes.SystemReservedEnforcementKey:
+		case kubetypes.KubeReservedEnforcementKey:
 			continue
 		default:
 			allErrors = append(allErrors, fmt.Errorf("Invalid option %q specified for EnforceNodeAllocatable (--enforce-node-allocatable) setting. Valid options are %q, %q or %q",
-				val, containermanager.NodeAllocatableEnforcementKey, containermanager.SystemReservedEnforcementKey, containermanager.KubeReservedEnforcementKey))
+				val, kubetypes.NodeAllocatableEnforcementKey, kubetypes.SystemReservedEnforcementKey, kubetypes.KubeReservedEnforcementKey))
 		}
 	}
 	return utilerrors.NewAggregate(allErrors)

--- a/pkg/kubelet/cm/BUILD
+++ b/pkg/kubelet/cm/BUILD
@@ -57,6 +57,7 @@ go_library(
             "//pkg/kubelet/events:go_default_library",
             "//pkg/kubelet/metrics:go_default_library",
             "//pkg/kubelet/qos:go_default_library",
+            "//pkg/kubelet/types:go_default_library",
             "//pkg/util/file:go_default_library",
             "//pkg/util/oom:go_default_library",
             "//pkg/util/procfs:go_default_library",

--- a/pkg/kubelet/cm/container_manager.go
+++ b/pkg/kubelet/cm/container_manager.go
@@ -122,13 +122,6 @@ type Status struct {
 	SoftRequirements error
 }
 
-const (
-	// Uer visible keys for managing node allocatable enforcement on the node.
-	NodeAllocatableEnforcementKey = "pods"
-	SystemReservedEnforcementKey  = "system-reserved"
-	KubeReservedEnforcementKey    = "kube-reserved"
-)
-
 // containerManager for the kubelet is currently an injected dependency.
 // We need to parse the --qos-reserve-requests option in
 // cmd/kubelet/app/server.go and there isn't really a good place to put

--- a/pkg/kubelet/cm/node_container_manager.go
+++ b/pkg/kubelet/cm/node_container_manager.go
@@ -32,6 +32,7 @@ import (
 	kubefeatures "k8s.io/kubernetes/pkg/features"
 	"k8s.io/kubernetes/pkg/kubelet/events"
 	evictionapi "k8s.io/kubernetes/pkg/kubelet/eviction/api"
+	kubetypes "k8s.io/kubernetes/pkg/kubelet/types"
 )
 
 const (
@@ -62,7 +63,7 @@ func (cm *containerManagerImpl) enforceNodeAllocatableCgroups() error {
 	// default cpu shares on cgroups are low and can cause cpu starvation.
 	nodeAllocatable := cm.capacity
 	// Use Node Allocatable limits instead of capacity if the user requested enforcing node allocatable.
-	if cm.CgroupsPerQOS && nc.EnforceNodeAllocatable.Has(NodeAllocatableEnforcementKey) {
+	if cm.CgroupsPerQOS && nc.EnforceNodeAllocatable.Has(kubetypes.NodeAllocatableEnforcementKey) {
 		nodeAllocatable = cm.getNodeAllocatableAbsolute()
 	}
 
@@ -101,7 +102,7 @@ func (cm *containerManagerImpl) enforceNodeAllocatableCgroups() error {
 		}()
 	}
 	// Now apply kube reserved and system reserved limits if required.
-	if nc.EnforceNodeAllocatable.Has(SystemReservedEnforcementKey) {
+	if nc.EnforceNodeAllocatable.Has(kubetypes.SystemReservedEnforcementKey) {
 		glog.V(2).Infof("Enforcing System reserved on cgroup %q with limits: %+v", nc.SystemReservedCgroupName, nc.SystemReserved)
 		if err := enforceExistingCgroup(cm.cgroupManager, nc.SystemReservedCgroupName, nc.SystemReserved); err != nil {
 			message := fmt.Sprintf("Failed to enforce System Reserved Cgroup Limits on %q: %v", nc.SystemReservedCgroupName, err)
@@ -110,7 +111,7 @@ func (cm *containerManagerImpl) enforceNodeAllocatableCgroups() error {
 		}
 		cm.recorder.Eventf(nodeRef, v1.EventTypeNormal, events.SuccessfulNodeAllocatableEnforcement, "Updated limits on system reserved cgroup %v", nc.SystemReservedCgroupName)
 	}
-	if nc.EnforceNodeAllocatable.Has(KubeReservedEnforcementKey) {
+	if nc.EnforceNodeAllocatable.Has(kubetypes.KubeReservedEnforcementKey) {
 		glog.V(2).Infof("Enforcing kube reserved on cgroup %q with limits: %+v", nc.KubeReservedCgroupName, nc.KubeReserved)
 		if err := enforceExistingCgroup(cm.cgroupManager, nc.KubeReservedCgroupName, nc.KubeReserved); err != nil {
 			message := fmt.Sprintf("Failed to enforce Kube Reserved Cgroup Limits on %q: %v", nc.KubeReservedCgroupName, err)

--- a/pkg/kubelet/eviction/BUILD
+++ b/pkg/kubelet/eviction/BUILD
@@ -18,7 +18,6 @@ go_test(
         "//pkg/apis/core:go_default_library",
         "//pkg/features:go_default_library",
         "//pkg/kubelet/apis/stats/v1alpha1:go_default_library",
-        "//pkg/kubelet/cm:go_default_library",
         "//pkg/kubelet/eviction/api:go_default_library",
         "//pkg/kubelet/lifecycle:go_default_library",
         "//pkg/kubelet/types:go_default_library",

--- a/pkg/kubelet/eviction/helpers.go
+++ b/pkg/kubelet/eviction/helpers.go
@@ -29,9 +29,9 @@ import (
 	utilfeature "k8s.io/apiserver/pkg/util/feature"
 	"k8s.io/kubernetes/pkg/features"
 	statsapi "k8s.io/kubernetes/pkg/kubelet/apis/stats/v1alpha1"
-	"k8s.io/kubernetes/pkg/kubelet/cm"
 	evictionapi "k8s.io/kubernetes/pkg/kubelet/eviction/api"
 	"k8s.io/kubernetes/pkg/kubelet/server/stats"
+	kubetypes "k8s.io/kubernetes/pkg/kubelet/types"
 	schedulerutils "k8s.io/kubernetes/plugin/pkg/scheduler/util"
 )
 
@@ -198,7 +198,7 @@ func parseThresholdStatement(signal evictionapi.Signal, val string) (evictionapi
 // getAllocatableThreshold returns the thresholds applicable for the allocatable configuration
 func getAllocatableThreshold(allocatableConfig []string) []evictionapi.Threshold {
 	for _, key := range allocatableConfig {
-		if key == cm.NodeAllocatableEnforcementKey {
+		if key == kubetypes.NodeAllocatableEnforcementKey {
 			return []evictionapi.Threshold{
 				{
 					Signal:   evictionapi.SignalAllocatableMemoryAvailable,

--- a/pkg/kubelet/eviction/helpers_test.go
+++ b/pkg/kubelet/eviction/helpers_test.go
@@ -30,8 +30,8 @@ import (
 	api "k8s.io/kubernetes/pkg/apis/core"
 	"k8s.io/kubernetes/pkg/features"
 	statsapi "k8s.io/kubernetes/pkg/kubelet/apis/stats/v1alpha1"
-	"k8s.io/kubernetes/pkg/kubelet/cm"
 	evictionapi "k8s.io/kubernetes/pkg/kubelet/eviction/api"
+	kubetypes "k8s.io/kubernetes/pkg/kubelet/types"
 	"k8s.io/kubernetes/pkg/quota"
 )
 
@@ -61,7 +61,7 @@ func TestParseThresholdConfig(t *testing.T) {
 			expectThresholds:        []evictionapi.Threshold{},
 		},
 		"all memory eviction values": {
-			allocatableConfig:       []string{cm.NodeAllocatableEnforcementKey},
+			allocatableConfig:       []string{kubetypes.NodeAllocatableEnforcementKey},
 			evictionHard:            map[string]string{"memory.available": "150Mi"},
 			evictionSoft:            map[string]string{"memory.available": "300Mi"},
 			evictionSoftGracePeriod: map[string]string{"memory.available": "30s"},

--- a/pkg/kubelet/types/constants.go
+++ b/pkg/kubelet/types/constants.go
@@ -24,4 +24,9 @@ const (
 	DockerContainerRuntime = "docker"
 	RktContainerRuntime    = "rkt"
 	RemoteContainerRuntime = "remote"
+
+	// User visible keys for managing node allocatable enforcement on the node.
+	NodeAllocatableEnforcementKey = "pods"
+	SystemReservedEnforcementKey  = "system-reserved"
+	KubeReservedEnforcementKey    = "kube-reserved"
 )

--- a/test/e2e_node/eviction_test.go
+++ b/test/e2e_node/eviction_test.go
@@ -28,8 +28,8 @@ import (
 	"k8s.io/kubernetes/pkg/features"
 	"k8s.io/kubernetes/pkg/kubelet/apis/kubeletconfig"
 	stats "k8s.io/kubernetes/pkg/kubelet/apis/stats/v1alpha1"
-	"k8s.io/kubernetes/pkg/kubelet/cm"
 	kubeletmetrics "k8s.io/kubernetes/pkg/kubelet/metrics"
+	kubetypes "k8s.io/kubernetes/pkg/kubelet/types"
 	"k8s.io/kubernetes/test/e2e/framework"
 
 	. "github.com/onsi/ginkgo"
@@ -104,7 +104,7 @@ var _ = framework.KubeDescribe("MemoryAllocatableEviction [Slow] [Serial] [Disru
 			initialConfig.KubeReserved = map[string]string{
 				string(v1.ResourceMemory): kubeReserved.String(),
 			}
-			initialConfig.EnforceNodeAllocatable = []string{cm.NodeAllocatableEnforcementKey}
+			initialConfig.EnforceNodeAllocatable = []string{kubetypes.NodeAllocatableEnforcementKey}
 			initialConfig.CgroupsPerQOS = true
 		})
 		runEvictionTest(f, pressureTimeout, expectedNodeCondition, logMemoryMetrics, []podEvictSpec{

